### PR TITLE
util-lib: rework get_process_cmdline()

### PR DESCRIPTION
**Pull request by poettering**
This reworks get_process_cmdline() quite substantially, fixing the following:
- Fixes:
  https://github.com/systemd/systemd/pull/3512/commits/a4e3bf4d7ac2de51191ce136ee9361ba319e106c#r66837630
- The passed max_length is also applied to the "comm" name, if comm_fallback is
  set.
- The right thing happens if max_length == 1 is specified
- when the cmdline "foobar" is abbreviated to 6 characters the result is now
  "foobar" instead of "foo...".
- trailing whitespace are removed before the ... suffix is appended. The 7
  character abbreviation of "foo barz" is hence "foo..." instead of "foo ...".
- leading whitespace are suppressed from the cmdline
- a comprehensive test case is added

### CODE REVIEW

File: src/test/test-process-util.c [hunk 1]
```diff
@@ -147,6 +155,170 @@ static void test_personality(void) {
...
+
+        assert_se(get_process_cmdline(getpid(), 7, true, &line) >= 0);
+        assert_se(streq(line, "[t...]"));
+        line = mfree(line);
+
+        assert_se(get_process_cmdline(getpid(), 8, true, &line) >= 0);
+        assert_se(streq(line, "[testa]"));
+        line = mfree(line);
+
+        assert_se(write(fd, "\0\0\0\0\0\0\0\0\0", 10) == 10);
```
Comment (mcspr) [inline src/test/test-process-util.c:227]: idk if this is just placeholder, but cmdline really could be just NULs.
max_length == 0 -> `len` would be 0 but line would still be allocated and returned
max_length > 1 -> `k` would be still equal to `r` and result in ENOENT

Comment (poettering) [inline src/test/test-process-util.c:227]: you mean if comm_fallback is false, then an entirely zero (but not empty) cmdline would result in the empty string being returned, but it should return -ENOENT?

Are you sure it should?

or maybe i am not grokking what you are saying?

Comment (mcspr) [inline src/test/test-process-util.c:227]: Yes, correct. I was just running dummy setproctitle process that zeroes argv and test-process-util (pid) fails instantly, because of comm_fallback=false at the start of testcases. But there is no empty string result

```
PID3032 comm: 'setproctitle'
PID3032 cmdline: '913 27745 34816 4913 4194304 95 0 0 0 0 0 0 0 20 0 1 0 314104177 8953856a
Assertion 'get_process_cmdline(pid, 8, false, &d) >= 0' failed at src/test/test-process-util.c:65, function test_get_process_comm(). Aborting.
Aborted
```

Comment (poettering) [inline src/test/test-process-util.c:227]: I think we should really treat an all-zero cmdline also as a reason to fall back to the comm field. Hence i think the current behaviour is actually right. That said, we should test this explicitly, hence I will post an updated patch shortly, that explicitly checks for this.

**Comment by poettering** [comment]
OK, I force pushed a new version that leaves the behaviour as before, but extends the test to verify this explicitly. This means we will fallback to comm whenever the cmdline is entirely absent or contains only zeroes, and is thus empty. I am pretty sure this is the right behaviour to follow here, as this call is primarily for presentational purposes and as such we should try to return something more useful than the empty string if we can.

I also added a longer comment to the function describing its behaviour, in particular some corner cases.

**Comment by evverx** [comment]
``` sh
$ sudo make check TESTS=test-process-util
[...]
FAIL: test-process-util
[...]

$ cat ./test-process-util.log
PID1 comm: 'systemd'
PID1 cmdline: '/usr/lib/systemd/systemd --system --deserialize 17'
PID1 cmdline truncated to 8: '/usr...0'
PID1 cmdline truncated to 1: ''
PID1 PPID: 0
PID1 exe: '/usr/lib/systemd/systemd'
PID1 UID: 0
PID1 GID: 0
PID1 strlen(environ): 0
PID1 $PATH: 'n/a'
PID29210 comm: 'test-process-ut'
PID29210 cmdline: './test-process-util'
PID29210 cmdline truncated to 8: './te...'
PID29210 cmdline truncated to 1: ''
PID29210 PPID: 29209
PID29210 exe: '/home/vagrant/systemd/test-process-util'
PID29210 UID: 0
PID29210 GID: 0
PID29210 strlen(environ): 2199
PID29210 $PATH: '/sbin:/bin:/usr/sbin:/usr/bin'
Assertion 'get_process_cmdline(getpid(), 0, false, &line) == -ENOENT' failed at src/test/test-process-util.c:229, function test_get_process_cmdline_harder(). Aborting.
Assertion 'si.si_code == CLD_EXITED' failed at src/test/test-process-util.c:173, function test_get_process_cmdline_harder(). Aborting.
FAIL test-process-util (exit status: 134)
```

`get_process_cmdline(getpid(), 0, false, &line)` returns `0`, `*line` contains some junk

``` sh
...
0 "[t...]"
test_get_process_cmdline_harder+0x73e [test-process-util]
main+0x85 [test-process-util]
__libc_start_main+0xf0 [libc-2.22.so]
_start+0x29 [test-process-util]
--

0 "[testa]"
test_get_process_cmdline_harder+0x7e0 [test-process-util]
main+0x85 [test-process-util]
__libc_start_main+0xf0 [libc-2.22.so]
_start+0x29 [test-process-util]
--

0 "\f$\255\373"
test_get_process_cmdline_harder+0x8c9 [test-process-util]
main+0x85 [test-process-util]
__libc_start_main+0xf0 [libc-2.22.so]
_start+0x29 [test-process-util]

ABRT
```

**Comment by evverx** [comment]
We allocate a memory [there](https://github.com/systemd/systemd/pull/3529/files#diff-4455c5d492b8b6c422f8c86c2929f861L123). But don't add trailing "\0".
So, [`isempty`](https://github.com/systemd/systemd/pull/3529/files#diff-4455c5d492b8b6c422f8c86c2929f861L185) returns `false` (sometimes)

Seems like `GREEDY_REALLOC0` fixes this

**Comment by keszybz** [comment]
Do we need the `if (len >0)` here https://github.com/systemd/systemd/pull/3529/files#diff-4455c5d492b8b6c422f8c86c2929f861R162? Maybe it should be `if (allocated > 0)` instead?

**Comment by poettering** [comment]
I prepped a fix for the issue @evverx pointed out in PR #3555, please have a look
